### PR TITLE
Refactor and rewrite with speedups and other improvements

### PR DESCRIPTION
**Pull request by aldanor**
Hi! I've spent a bit of time on reworking this crate (as I surprisingly found it to be the slowest part in the processing toolchain which got me to dig a bit deeper to figure out why's that).

Since there's so many changes and it's pretty much a backwards-incompatible rewrite (although the tests are kept and a few more are added), here's a few options, all of which are fine with me and understandable:

- This gets merged and you keep on being a sole maintainer (or add me as comaintainer if you want)
- You aren't able or don't wish to maintain this crate any longer and want to pass it on along with its name
- Something inbetween the previous two options where the maintenance and rights are shared in some way, whichever way you'd prefer
- You don't want to merge this and want to retain your crate, in which case I would have to publish another one since I need it working and published asap

Here's a list of some of the changes:

- Parsing algorithm is rewritten from scratch and is much faster
- Sequence mask replacing algorithm is rewritten from scratch and is much faster as well
- Docs are improved, many docstrings are rewritten
- Reader is no longer based on `BufReader<File>` only, but can be placed on top of any reader
- Built-in support for in-memory-cached files which provides lightning-fast decoding
- Remove half of the API by using Rust ranges everywhere to slice sequences when decoding
- Many other API changes, making it backwards-incompatible but nicer (in my opinion) and more Rust-like (but less py2bit-like, whether that's a good or bad thing)
- The entire codebase is now conforming to the latest clippy
- Added GitHub actions workflows so that the tests/clippy run automatically
- Added simple benchmarks that can be run directly via cargo

Timings comparison:

```
master:
    sequence                time:   [20.588 us 20.774 us 20.976 us]
    full_sequence           time:   [21.221 us 21.339 us 21.464 us]
PR (from disk):
    sequence                time:   [1.5980 us 1.6061 us 1.6150 us] (~13x)
    full_sequence           time:   [1.6980 us 1.7057 us 1.7137 us]
PR (from memory):
    sequence                time:   [105.97 ns 106.27 ns 106.62 ns] (~196x)
    full_sequence           time:   [171.51 ns 178.67 ns 185.83 ns]
```

### CODE REVIEW

**Comment by jbethune** [comment]
Wow. Thanks for the pull request! Looks promising at first sight. But I will need a bit of time to review everything and make the decisions you mentioned.

**Comment by aldanor** [comment]
> Wow. Thanks for the pull request! Looks promising at first sight. But I will need a bit of time to review everything and make the decisions you mentioned.

Sounds good - let me know if you have any questions or if there's anything to discuss.

File: Cargo.toml
Comment (jbethune) [inline Cargo.toml]: we need to bump the minor version because of API changes

Comment (aldanor) [reply Cargo.toml]: Agreed, needs to be changed to 0.2.0

File: src/block.rs
Comment (jbethune) [inline src/block.rs]: I personally would call the function just `overlaps`. I don't see any functions in the stdlib that are called `iter_something` (apart from  `iter_mut`). But I would also be fine with keeping it as it currently is.

File: src/block.rs
Comment (jbethune) [inline src/block.rs]: There should be a comment above this line that says something like `// safe because for-loop ensures valid bounds `

Comment (jbethune) [reply src/block.rs]: I actually doubt that this is really safe. The parameter `seq` does not get cross-checked by anything.

File: src/block.rs
Comment (aldanor) [reply src/block.rs]: Yep, agreed. In stdlib it's all like `keys()`, `values()`, etc, so `overlaps()` is more rustic.

File: src/block.rs
Comment (aldanor) [reply src/block.rs]: I think the way to make it safe is to just pass `start` instead of `range`.

Because the implicit invariant here now is that `seq.len() == range.end - range.start`.

File: src/lib.rs
Comment (jbethune) [inline src/lib.rs:192]: Out of curiosity: why the `const`?

**Comment by aldanor** [comment]
@jbethune I've rewritten `apply_masks()` with no unsafe and in this case it's not slower (but actually a few percent faster, at least on my machine) - so, good call!

File: src/lib.rs
Comment (aldanor) [reply src/lib.rs:192]: Ah, that's to shut clippy up - it asks you to make things const if they can be const. Otherwise you'd have to either add a crate-wide lint ignore (which isn't nice because in some cases const methods may make sense), or add a method-specific lint ignore (which looks worse than just adding a const).

File: src/lib.rs
Comment (jbethune) [inline src/lib.rs:303]: A typecast to underscore? Is this type erasure like a `(void*)` pointer in C?

File: src/lib.rs
Comment (jbethune) [inline src/lib.rs:310]: Needs a comment above the unsafe block.

Comment (jbethune) [reply src/lib.rs:310]: Maybe we can actually use an array instead of a vec here? Then the size of the array would be constant.

File: src/lib.rs
Comment (jbethune) [inline src/lib.rs]: Comment about why this is safe

File: src/lib.rs
Comment (aldanor) [reply src/lib.rs:310]: We can't use an array here because length is not known at compile time

File: src/lib.rs
Comment (aldanor) [reply src/lib.rs:303]: No, it's not type erasure - it's basically when the type of the destination is known, you don't have to type it explicitly when using `as`, and can just use an underscore. E.g. `let x: i64 = 10_i32 as _`;

File: src/lib.rs
Comment (aldanor) [reply src/lib.rs]: I'll check whether something like `[u8]::chunk_exact()` can be used here without performance degradation. Otherwise I'll add a comment.

File: src/lib.rs
Comment (jbethune) [reply src/lib.rs:310]: Makes sense. But maybe we could do this instead?

`let mut out = vec![0; length];`

I created a simple Rust program that included that line and I looked at the generated assembly code. The code calls `memset` which is very efficient at initializing a region of memory.

**Comment by jbethune** [comment]
Congratulations! You are hereby a Co-Maintainer. I really like what you have done with the code!

File: src/lib.rs
Comment (aldanor) [reply src/lib.rs:310]: It's not a big deal here, but I would argue that simple `malloc` is still a bit faster - why set all those zeroes when we don't need it as they will get overwritten a few lines below? (especially when the sequence is long)

`set_len()` is unsafe in two cases:
- capacity is smaller than length
- objects inside the vector require some invariants to hold, or are non-trivial

Here, none of the above is a problem - we just allocated enough capacity for resizing, and there are no invariants attached to simple integers, whatever they're equal to is fine and they will get overwritten anyway.

**Comment by aldanor** [comment]
> Congratulations! You are hereby a Co-Maintainer. I really like what you have done with the code!

Thanks! Always glad to participate 👍

**Comment by aldanor** [comment]
Just a note: I've received an email re: crate co-ownership from https://crates.io, but it seems like the invitation has expired somehow, or something went wrong, maybe try again?

**Comment by aldanor** [comment]
@jbethune I've rewritten the decoder once again 🤦 (see 177d890)

- On a local machine, went from 119ns to 98ns in the partial bench (approximately 20% speedup)
- No more unsafe in the main loop
- Much less code
- No more bitshifts, no more math
- Uses a preparsed "database" of quadruples (see 734575c)

**Comment by jbethune** [comment]
> Just a note: I've received an email re: crate co-ownership from https://crates.io, but it seems like the invitation has expired somehow, or something went wrong, maybe try again?

Is it still not working? [You are listed in the owners section](https://crates.io/crates/twobit)

**Comment by aldanor** [comment]
> Neat! It is such a joy to find new toys in the standard library to simplify code. Well done!

Agreed - I've never used `chunks` API before (and btw - there's new `as_chunks` that uses const generics and will be stabilized at some point, that looks quite interesting).

The speedup though is mainly due to preparsing quadruples, I think.

**Comment by aldanor** [comment]
> > Just a note: I've received an email re: crate co-ownership from https://crates.io, but it seems like the invitation has expired somehow, or something went wrong, maybe try again?
> 
> Is it still not working? [You are listed in the owner's section](https://crates.io/crates/twobit)

I think it works now, maybe something was lagging yesterday and it didn't update yet back then.

So what's the plan then? 

- I'll add a brief changelog, it's a good thing to have.
- If everything looks fine, I can try merging this in (and we need to ensure that GitHub Actions will then kick off in this repo and start running tests, first of all)
- Add 0.2.0 git tag and cargo publish that commit, then add a github release (by the way - I can't seem to edit description of the github repo etc - perhaps there's some more checkboxes you might have to set in the repository settings if you want to)
- One more thing - would you like the license to be MIT or MIT + Apache2?

**Review by jbethune — APPROVED** [review]
> by the way - I can't seem to edit description of the github repo etc - perhaps there's some more checkboxes you might have to set in the repository settings if you want to

I did not find anything in the repo options. But I found [this stackoverflow thread)[https://stackoverflow.com/questions/7757751/how-do-you-change-a-repository-description-on-github) and I guess that you will be able to edit the description, once this pull request has been merged.

I'd like to stick to just the MIT license for now. I agree with the remaining suggestions you have made.

**Review by jbethune — APPROVED** [review]
> by the way - I can't seem to edit description of the github repo etc - perhaps there's some more checkboxes you might have to set in the repository settings if you want to

I did not find anything in the repo options. But I found [this stackoverflow thread](https://stackoverflow.com/questions/7757751/how-do-you-change-a-repository-description-on-github) and I guess that you will be able to edit the description, once this pull request has been merged.

I'd like to stick to just the MIT license for now. I agree with the remaining suggestions you have made.

**Comment by jbethune** [comment]
You can merge any time you feel ready.

**Pull request merged**
